auction-server/src/kernel/rpc_client_svm_tester.rs

```diff
@@ -0,0 +1,208 @@
+        nonblocking::rpc_client::RpcClient,
+        rpc_client::RpcClientConfig,
+        rpc_sender::{
+            RpcSender,
+            RpcTransportStats,
+        },
+    },
+    solana_sdk::account::Account,
+    std::sync::Arc,
+    tokio::sync::Mutex,
+};
+
+pub enum CannedRequestMatcher {
+    AllByRequest(RpcRequest),
+    MatchRequestAndParams(RpcRequest, serde_json::Value),
+    MatchRequestAndParamsDynamically(
+        RpcRequest,
+        Box<dyn Fn(&serde_json::Value) -> bool + Send + Sync>,
+    ),
+}
+
+impl CannedRequestMatcher {
+    pub fn is_match(&self, request: &RpcRequest, params: &serde_json::Value) -> bool {
+        match self {
+            CannedRequestMatcher::AllByRequest(req) => req == request,
+            CannedRequestMatcher::MatchRequestAndParams(req, p) => req == request && p == params,
+            CannedRequestMatcher::MatchRequestAndParamsDynamically(req, f) => {
+                req == request && f(params)
+            }
+        }
+    }
+}
+
+pub type DynamicResultFn =
+    Box<dyn Fn(&RpcRequest, serde_json::Value) -> Result<serde_json::Value> + Send + Sync>;
+
+pub enum CannedResult {
+    Static(serde_json::Value),
+    DynamicByParams(DynamicResultFn),
+}
+
+impl CannedResult {
+    pub fn get_result(
+        &self,
+        request: &RpcRequest,
+        params: &serde_json::Value,
+    ) -> Result<serde_json::Value> {
+        match self {
+            CannedResult::Static(r) => Ok(r.clone()),
+            CannedResult::DynamicByParams(f) => f(request, params.clone()),
+        }
+    }
+}
+
+pub struct CannedRequest {
+    matcher:  CannedRequestMatcher,
+    result:   CannedResult,
+    hits:     usize,
+    max_hits: Option<usize>,
+}
+
+impl CannedRequest {
+    pub fn new(
+        matcher: CannedRequestMatcher,
+        result: CannedResult,
+        max_hits: Option<usize>,
+    ) -> Self {
+        Self {
+            matcher,
+            result,
+            hits: 0,
+            max_hits,
+        }
+    }
+
+    pub fn is_dead(&self) -> bool {
+        self.max_hits.map(|max| self.hits >= max).unwrap_or(false)
+    }
+}
+
+pub struct RpcClientSvmTesterInner {
+    canned_responses: Mutex<Vec<CannedRequest>>,
+}
+
+#[derive(Clone)]
+pub struct RpcClientSvmTester(Arc<RpcClientSvmTesterInner>);
+
+impl std::ops::Deref for RpcClientSvmTester {
+    type Target = RpcClientSvmTesterInner;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl RpcClientSvmTester {
+    pub fn new() -> Self {
+        Self(Arc::new(RpcClientSvmTesterInner {
+            canned_responses: Mutex::new(vec![]),
+        }))
+    }
+
+    pub fn make_test_client(&self) -> RpcClient {
+        RpcClient::new_sender(
+            RpcClientTesterSender {
+                tester: self.clone(),
+            },
+            RpcClientConfig::default(),
+        )
+    }
+
+    /// Panics if not all the canned responses have been consumed during the test
+    pub async fn check_all_uncanned(&self) {
```

Comment:
could we do this at drop?

Reply:
its async, async drop last time i checked wasnt too nice

Reply:
yeah there is just nothing really to poll once u drop i suppose